### PR TITLE
SSL / BlockHound works out of the box with the default SSL provider

### PR DESCRIPTION
**Pull request by johnou**
Motivation:

JDK is the default SSL provider and internally uses blocking IO operations.

Modifications:

Add allowBlockingCallsInside configuration for SslHandler handshake function.

Result:

When BlockHound is installed, SSL works out of the box with the default SSL provider.

### CODE REVIEW

**Comment by violetagg** [comment]
@johnou Can we add also a test like this or similar https://github.com/violetagg/netty/commit/e24767c5d894978b0861e3e2a104395cb45a97a3

**Comment by johnou** [comment]
@violetagg of course, is it ok if I use your commit? have you already signed the ICLA?

**Comment by violetagg** [comment]
> is it ok if I use your commit? 

Yep it is OK

> have you already signed the ICLA?

Yep (I do have other contributions for Netty project)

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport-blockhound-tests/src/test/java/io/netty/util/internal/NettyBlockHoundIntegrationTest.java [hunk 1]
```diff
@@ -66,4 +92,89 @@ public void testBlockingCallsInNettyThreads() throws Exception {
...
+    public void testHandshakeWithExecutor() throws Exception {
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        try {
+            testHandshakeWithExecutor(executorService);
+        } finally {
+            executorService.shutdown();
+        }
+    }
+
+    private void testHandshakeWithExecutor(Executor executor) throws Exception {
```
Comment (normanmaurer) [inline transport-blockhound-tests/src/test/java/io/netty/util/internal/NettyBlockHoundIntegrationTest.java]: nit: static ?

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by johnou** [comment]
from CI..

```
Suppressed: java.lang.IllegalAccessError: class io.netty.handler.ssl.util.OpenJdkSelfSignedCertGenerator (in unnamed module @0x7642df8f) cannot access class sun.security.x509.X509CertInfo (in module java.base) because module java.base does not export sun.security.x509 to unnamed module @0x7642df8f
		at io.netty.handler.ssl.util.OpenJdkSelfSignedCertGenerator.generate(OpenJdkSelfSignedCertGenerator.java:52)
		at io.netty.handler.ssl.util.SelfSignedCertificate.<init>(SelfSignedCertificate.java:158)
		... 31 more
Caused by: java.lang.NoClassDefFoundError: org/bouncycastle/cert/X509v3CertificateBuilder
	at io.netty.handler.ssl.util.SelfSignedCertificate.<init>(SelfSignedCertificate.java:163)
	... 31 more
Caused by: java.lang.ClassNotFoundException: org.bouncycastle.cert.X509v3CertificateBuilder
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 32 more
```

@normanmaurer I added bouncycastle to the blockhound pom, can we run tests again?

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by johnou** [comment]
@violetagg the test is failing on java11+ (sun.security.x509 is not exported) so it immediately falls back to bouncycastle which doesn't use blocking IO :)

**Comment by johnou** [comment]
@normanmaurer I suppose we could add `--add-exports java.base/sun.security.x509=ALL-UNNAMED` to argLine.java9 for Java10 profile?

**Comment by violetagg** [comment]
at least that was done for `handler` module

https://github.com/netty/netty/blob/4.1/handler/pom.xml#L31-L32

**Comment by normanmaurer** [comment]
@johnou yes please do

**Comment by johnou** [comment]
hopefully in the right spot, I don't have jdk11 on this rig.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by johnou** [comment]
@violetagg tests still failing on java11+ https://ci.netty.io/job/netty-centos6-java12-prb/1756/testReport/junit/io.netty.util.internal/NettyBlockHoundIntegrationTest/testHandshakeWithImmediateEventExecutor/

**Comment by violetagg** [comment]
@johnou checking

**Comment by violetagg** [comment]
just an update: testing on my side on `centos-8-v20191210` with `openjdk version "11.0.5" 2019-10-15 LTS` and cannot reproduce
is there anything specific to these jobs?

**Comment by violetagg** [comment]
@normanmaurer @johnou Even though that I cannot reproduce it. From what I see as an output from the test - the work is not offloaded. The corresponding code and stack trace are:

https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1506-L1509

https://ci.netty.io/job/netty-centos6-java11-prb/1997/consoleText

```
15:19:26.718 [nioEventLoopGroup-4-2] DEBUG i.n.h.s.u.InsecureTrustManagerFactory - Accepting a server certificate: CN=example.com
java.lang.Error: Blocking call! java.io.FileInputStream#readBytes
	at reactor.blockhound.BlockHound$Builder.lambda$new$0(BlockHound.java:196)
	at reactor.blockhound.BlockHound$Builder.lambda$install$6(BlockHound.java:318)
	at reactor.blockhound.BlockHoundRuntime.checkBlocking(BlockHoundRuntime.java:46)
	at java.base/java.io.FileInputStream.readBytes(FileInputStream.java)
	at java.base/java.io.FileInputStream.read(FileInputStream.java:279)
	at java.base/java.io.FilterInputStream.read(FilterInputStream.java:133)
	at java.base/sun.security.provider.NativePRNG$RandomIO.readFully(NativePRNG.java:424)
	at java.base/sun.security.provider.NativePRNG$RandomIO.ensureBufferValid(NativePRNG.java:526)
	at java.base/sun.security.provider.NativePRNG$RandomIO.implNextBytes(NativePRNG.java:545)
	at java.base/sun.security.provider.NativePRNG.engineNextBytes(NativePRNG.java:220)
	at java.base/java.security.SecureRandom.nextBytes(SecureRandom.java:741)
	at jdk.crypto.ec/sun.security.ec.ECKeyPairGenerator.generatePrivateScalar(ECKeyPairGenerator.java:168)
	at jdk.crypto.ec/sun.security.ec.ECKeyPairGenerator.generateKeyPairImpl(ECKeyPairGenerator.java:194)
	at jdk.crypto.ec/sun.security.ec.ECKeyPairGenerator.generateKeyPair(ECKeyPairGenerator.java:150)
	at java.base/java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:728)
	at java.base/sun.security.ssl.ECDHKeyExchange$ECDHEPossession.<init>(ECDHKeyExchange.java:128)
	at java.base/sun.security.ssl.ECDHClientKeyExchange$ECDHEClientKeyExchangeProducer.produce(ECDHClientKeyExchange.java:390)
	at java.base/sun.security.ssl.ClientKeyExchange$ClientKeyExchangeProducer.produce(ClientKeyExchange.java:65)
	at java.base/sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:436)
	at java.base/sun.security.ssl.ServerHelloDone$ServerHelloDoneConsumer.consume(ServerHelloDone.java:182)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1061)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1048)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:995)
	at io.netty.handler.ssl.SslHandler.runAllDelegatedTasks(SslHandler.java:1494)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1508)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1392)
```

**Comment by normanmaurer** [comment]
@johnou I think the correct thing would be to whitelist `SslHandler.runAllDelegatedTasks`

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@johnou @violetagg thanks a lot!

**Comment by violetagg** [comment]
thanks

**Pull request merged**